pkg/metrics/metrics.go

```diff
@@ -0,0 +1,38 @@
+package metrics
+
+import "github.com/prometheus/client_golang/prometheus"
+
+const MetricsNamespace = "adapter"
+
+var PrometheusUp = prometheus.NewGauge(prometheus.GaugeOpts{
+	Namespace: MetricsNamespace,
+	Name:      "prometheus_up",
+	Help:      "1 when adapter is able to reach prometheus, 0 otherwise",
+})
+
+var RegistryMetrics = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+	Namespace: MetricsNamespace,
+	Name:      "registry_metrics",
+	Help:      "number of metrics entries in cache registry",
+}, []string{"registry"})
+
+var Errors = prometheus.NewCounterVec(prometheus.CounterOpts{
+	Namespace: MetricsNamespace,
+	Name:      "errors_total",
+	Help:      "number of errors served",
+}, []string{"type"})
+
+var Rules = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+	Namespace: MetricsNamespace,
+	Name:      "roles",
+	Help:      "number of configured rules",
+}, []string{"type"})
+
+func init() {
+	prometheus.MustRegister(
```

Comment:
Let's not use `init()` and register the prometheus registry as part of package initialization. Let's rather intanciate the registry in `main.go`, and dependency inject it into the providers.

Reply:
I feel _this_ is non-idiomatic for the Prometheus client SDK, but you're the boss. :)

It's going to be more complicated than just injecting a registry, because the registry doesn't actually provide access to the metrics registered to it. We'll have to inject either the metric of interest, or an object that wraps all of the metrics.

Reply:
Injecting the metric of interest sounds good to me :+1: 

Regarding dependency injection: Indeed, historically `init()` based registry initialization has been used a lot within the Go ecosystem but many Prometheus ecosystem maintainers recommend dependency injection (be it the metric or the registry itself) instead. Also see the metrics overhaul KEP [1] for the motivation of not using `init()` based metrics registration.

[1] https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20181106-kubernetes-metrics-overhaul.md#motivation

Reply:
yep, all good.

I'm happy to rework to inject just the metric, but I _did_ already push commits that refactor to inject a ServiceMetrics object. Are you OK with that approach, or do you feel strongly that it should only be prometheus' Gauge (etc.) object injected?